pkg/emb6/contrib/sock/udp/emb6_sock_udp.c

```diff
@@ -0,0 +1,323 @@
+/*
+ * Copyright (C) 2016 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @{
+ *
+ * @file
+ * @author  Martine Lenders <m.lenders@fu-berlin.de>
+ */
+
+#include <assert.h>
+#include <errno.h>
+#include <stdbool.h>
+
+#include "byteorder.h"
+#include "evproc.h"
+#include "msg.h"
+#include "mutex.h"
+#include "net/af.h"
+#include "net/sock/udp.h"
+#include "net/ipv6/hdr.h"
+#include "sched.h"
+#include "uip.h"
+#include "xtimer.h"
+
+#define _MSG_TYPE_CLOSE     (0x4123)
+#define _MSG_TYPE_TIMEOUT   (0x4124)
+#define _MSG_TYPE_RCV       (0x4125)
+
+/* struct to describe a sendto command for emb6 thread */
+typedef struct {
+    mutex_t block;
+    struct udp_socket *sock;
+    const sock_udp_ep_t *remote;
+    int res;
+    const void *data;
+    size_t len;
+} _send_cmd_t;
+
+extern uint16_t uip_slen;
+
+static bool send_registered = false;
+
+static void _timeout_callback(void *arg);
+static void _input_callback(struct udp_socket *c, void *ptr,
+                            const uip_ipaddr_t *src_addr, uint16_t src_port,
+                            const uip_ipaddr_t *dst_addr, uint16_t dst_port,
+                            const uint8_t *data, uint16_t datalen);
+static void _output_callback(c_event_t c_event, p_data_t p_data);
+
+static int _reg(struct udp_socket *c, void *ptr, udp_socket_input_callback_t cb,
+                const sock_udp_ep_t *local, const sock_udp_ep_t *remote)
+{
+    if (((local != NULL) && (local->family != AF_INET6)) ||
+        ((remote != NULL) && (remote->family != AF_INET6))) {
+        return -EAFNOSUPPORT;
+    }
+    if (udp_socket_register(c, ptr, cb) < 0) {
+        return -ENOMEM;
+    }
+    if (local != NULL) {
+        if (udp_socket_bind(c, local->port) < 0) {
+            udp_socket_close(c);
+            return -EADDRINUSE;
+        }
+    }
+    if (remote != NULL) {
+        udp_socket_connect(c, (uip_ipaddr_t *)&remote->addr, remote->port);
```

Comment:
mhm, I think that is exactly what happens below in L243 where `_reg` is called with `&tmp` and `tmp->udp_conn` is never set, hence NULL or something random.